File: python/argus_utils/MatrixUtils.py
#!/usr/bin/env python

import numpy as np
from argus_msgs.msg import MatrixFloat64

def MsgToMatrix( msg ):
    if len( msg.data != msg.rows * msg.cols ):
        raise RuntimeError( 'Message matrix dimensions do not match data.' )

    if msg.column_major:
        return np.reshape( msg.data, [msg.rows, msg.cols], order='F' )
    else:
        return np.reshape( msg.data, [msg.rows, msg.cols], order='C' )

def MatrixToMsg( mat ):
    if len(mat.shape) > 2:
        raise RuntimeError( 'Cannot convert array with more than 2 dimensions.' )

    msg = MatrixFloat64()
    msg.column_major = True;
    msg.rows = mat.shape[0]
    msg.cols = mat.shape[1]
    msg.data = mat.flatten( order='F' )
    return msg

File: argus_utils/src/MatrixUtils.cpp
#include "argus_utils/utils/MatrixUtils.h"

namespace argus
{

MatrixType MsgToMatrix( const argus_msgs::MatrixFloat64& msg )
{
	if( msg.rows*msg.cols != msg.data.size() )
	{
		throw std::runtime_error( "MsgToMatrix: Incorrect sized matrix message." );
	}

	if( msg.column_major )
	{
		return Eigen::Map<const ColMatrixType>( msg.data.data(), msg.rows, msg.cols );
	}
	else
	{
		return Eigen::Map<const RowMatrixType>( msg.data.data(), msg.rows, msg.cols );
	}
}

argus_msgs::MatrixFloat64 MatrixToMsg( const MatrixType& mat )
{
	argus_msgs::MatrixFloat64 msg;
	msg.column_major = true;
	msg.rows = mat.rows();
	msg.cols = mat.cols();
	SerializeMatrix( mat, msg.data, ColMajor );
	return msg;
}

MatrixType MsgToSymmetric( const argus_msgs::SymmetricFloat64& msg )
{
	if( msg.dim * msg.dim != msg.data.size() )
	{
		throw std::runtime_error( "MsgToSymmetric: Incorrect sized matrix message." );
	}

	// Row and column major don't matter for symmetric
	MatrixType mat( msg.dim, msg.dim );
	unsigned int ind = 0;
	for( unsigned int i = 0; i < msg.dim; ++i )
	{
		for( unsigned int j = i; j < msg.dim; ++j )
		{
			mat(i,j) = msg.data[ind];
			mat(j,i) = msg.data[ind];
			++ind;
		}
	}
	return mat;
}

argus_msgs::SymmetricFloat64 SymmetricToMsg( const MatrixType& mat )
{
	if( mat.rows() != mat.cols() )
	{
		throw std::runtime_error( "SymmetricToMsg: Matrix not square." );
	}

	argus_msgs::SymmetricFloat64 msg;
	msg.dim = mat.rows();
	msg.data.reserve( msg.dim*(msg.dim+1)/2 );
	for( unsigned int i = 0; i < msg.dim; ++i )
	{
		for( unsigned int j = i; j < msg.dim; ++j )
		{
			msg.data.push_back( mat(i,j) );
		}
	}
	return msg;
}


}